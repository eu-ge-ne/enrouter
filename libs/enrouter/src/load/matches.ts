import { createLog } from "#log.js";

import { loaders } from "./loaders.js";

import type { RouteModules } from "#modules.js";
import type { RouteMatch } from "#matches/mod.js";

const log = createLog("load/matches");

interface LoadRouteMatchesParams {
  matches: RouteMatch[];
  modules: RouteModules;
}

export async function loadRouteMatches({
  matches,
  modules,
}: LoadRouteMatchesParams): Promise<void> {
  log("Loading route matches");

  const promises: Promise<unknown>[] = [];

  for (const { handler } of matches) {
    for (const module of handler.modules) {
      if (module.loaded) {
        continue;
      }

      const { path, load } = modules[module.id]!;

      const fileName = path.slice(path.lastIndexOf("/") + 1);

      const promise = loaders[fileName]?.({ handler, module, load });
      if (promise) {
        promises.push(promise);
      }
    }
  }

  await Promise.all(promises);

  log("Route matches loaded");
}
