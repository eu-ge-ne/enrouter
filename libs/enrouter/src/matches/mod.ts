import { createLog } from "#log.js";

import type { RouteHandler } from "#handlers/mod.js";

const log = createLog("matches");

export interface RouteMatch {
  handler: RouteHandler;
  location: string;
  params: Record<string, string>;
  next?: RouteMatch;
}

interface MatchRoutesParams {
  handlers: RouteHandler;
  location: string;
}

export function matchRoutes({
  handlers,
  location,
}: MatchRoutesParams): RouteMatch[] {
  log("Matching routes for location %s", location);

  const matches: RouteMatch[] = [];

  recur([handlers], location, matches);

  for (let i = 0; i < matches.length; i += 1) {
    const match = matches[i]!;
    match.next = matches[i + 1];
  }

  log("Matched routes %O", matches);

  return matches;
}

function recur(
  handlers: RouteHandler[],
  location: string,
  matches: RouteMatch[],
): void {
  let matchedHandlers = handlers
    .map((x) => [x, x.test?.pattern.exec(location)] as const)
    .filter((x): x is [RouteHandler, RegExpExecArray] => Boolean(x[1]));

  if (matchedHandlers.length === 0) {
    return;
  }

  // TODO: improve
  if (matchedHandlers.length > 1) {
    matchedHandlers = matchedHandlers.filter(
      (x) => x[0].test?.keys.length === 0,
    );
  }

  const [handler, results] = matchedHandlers[0]!;

  const params = Object.fromEntries(
    handler.test.keys.map((key, i) => {
      return [key, results![i + 1]!] as const;
    }),
  );

  matches.push({ handler, location: results[0] || "/", params });

  if (handler.tree) {
    recur(handler.tree, location, matches);
  }
}
