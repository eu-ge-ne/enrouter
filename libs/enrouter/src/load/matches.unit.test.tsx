import { describe, test, expect } from "vitest";
import * as regexparam from "regexparam";

import { loadRouteMatches } from "./matches.js";

import type { RouteModules } from "#modules.js";
import type { RouteMatch } from "#matches/mod.js";

describe("loadRouteMatches", () => {
  test("no matches", async () => {
    const matches: RouteMatch[] = [];

    await loadRouteMatches({ matches, modules: {} });

    expect(matches).toMatchSnapshot();
  });

  test("1 match with no modules", async () => {
    const matches: RouteMatch[] = [
      {
        handler: {
          route: {
            path: "/",
            mod: [],
            link: [[], []],
          },
          test: regexparam.parse("/", true),
          modules: [],
        },
        location: "/",
        params: {},
      },
    ];

    await loadRouteMatches({ matches, modules: {} });

    expect(matches).toMatchSnapshot();
  });

  test("1 match with _layout.tsx module", async () => {
    const matches: RouteMatch[] = [
      {
        handler: {
          route: {
            path: "/",
            mod: ["_layout.tsx"],
            link: [[], []],
          },
          test: regexparam.parse("/", true),
          modules: [{ id: "_layout.tsx" }],
        },
        location: "/",
        params: {},
      },
    ];

    const modules: RouteModules = {
      "_layout.tsx": {
        path: "src/_layout.tsx",
        load: async () => ({
          components: {
            main: () => <div>Layout</div>,
          },
        }),
      },
    };

    await loadRouteMatches({ matches, modules });

    expect(matches).toMatchSnapshot();
  });

  test("1 match with _index.tsx module", async () => {
    const matches: RouteMatch[] = [
      {
        handler: {
          route: {
            path: "/",
            mod: ["_index.tsx"],
            link: [[], []],
          },
          test: regexparam.parse("/", true),
          modules: [{ id: "_index.tsx" }],
        },
        location: "/",
        params: {},
      },
    ];

    const modules: RouteModules = {
      "_index.tsx": {
        path: "src/_index.tsx",
        load: async () => ({
          components: {
            main: () => <div>Index</div>,
          },
        }),
      },
    };

    await loadRouteMatches({ matches, modules });

    expect(matches).toMatchSnapshot();
  });
});
