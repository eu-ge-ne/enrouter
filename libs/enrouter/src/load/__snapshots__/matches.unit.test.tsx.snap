// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadRouteMatches > 1 match with _index.tsx module 1`] = `
[
  {
    "handler": {
      "index": {
        "main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
      "modules": [
        {
          "id": "_index.tsx",
          "loaded": true,
        },
      ],
      "route": {
        "link": {
          "css": [],
          "mod": [],
        },
        "mod": [
          "_index.tsx",
        ],
        "path": "/",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
    "params": {},
  },
]
`;

exports[`loadRouteMatches > 1 match with _layout.tsx module 1`] = `
[
  {
    "handler": {
      "layout": {
        "main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
      "modules": [
        {
          "id": "_layout.tsx",
          "loaded": true,
        },
      ],
      "route": {
        "link": {
          "css": [],
          "mod": [],
        },
        "mod": [
          "_layout.tsx",
        ],
        "path": "/",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
    "params": {},
  },
]
`;

exports[`loadRouteMatches > 1 match with no modules 1`] = `
[
  {
    "handler": {
      "modules": [],
      "route": {
        "link": {
          "css": [],
          "mod": [],
        },
        "mod": [],
        "path": "/",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
    "params": {},
  },
]
`;

exports[`loadRouteMatches > no matches 1`] = `[]`;
